### PR TITLE
Add cart logic

### PR DESCRIPTION
**Pull request by artemklishch**
*Made cart controller
  *Made cart mode
  *Made cart item model
  *Made liquibase settings for creating tables for carts and cart items
  *Made cart and cart items repositries
  *Made cart service
  *Made cart and cart item mappers
  *Updated the book mapper
  *Made dto`s

### CODE REVIEW

File: src/main/java/org/example/intro/controller/CartController.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+    @GetMapping
+    @PreAuthorize("hasAnyRole('USER')")
+    @Operation(
+            summary = "Retrieve the shopping cart",
+            description = "Retrieve the shopping cart data with cart items"
+    )
+    public ShoppingCartDto getCart(
+            Authentication authentication,
+            Pageable pageable
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/controller/CartController.java]:
```suggestion
```

File: src/main/java/org/example/intro/controller/CartController.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    @PostMapping
+    @PreAuthorize("hasAnyRole('USER')")
+    @ResponseStatus(HttpStatus.CREATED)
+    @Operation(
+            summary = "Add cart item to the shopping cart",
+            description = "Add cart item to the shopping cart"
+    )
+    public CartItemDto addToCart(
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/controller/CartController.java]:
```suggestion
    public ShoppingCartDto addToCart(
```

File: src/main/java/org/example/intro/controller/CartController.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    @PutMapping("/items/{cartItemId}")
+    @PreAuthorize("hasAnyRole('USER')")
+    @ResponseStatus(HttpStatus.OK)
+    @Operation(
+            summary = "Update cart item quantity",
+            description = "Update cart item quantity in the shopping cart"
+    )
+    public CartItemDto updateCartItemsQuantity(
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/controller/CartController.java]:
```suggestion
    public ShoppingCartDto updateCartItemsQuantity(
```

File: src/main/java/org/example/intro/controller/CartController.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+            @PathVariable Long cartItemId,
+            @RequestBody UpdateCartItemsQuantityDto quantityDto
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/controller/CartController.java]:
```suggestion
            @RequestBody @Valid UpdateCartItemsQuantityDto quantityDto
```

File: src/main/java/org/example/intro/controller/CartController.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    @PostMapping
+    @PreAuthorize("hasAnyRole('USER')")
+    @ResponseStatus(HttpStatus.CREATED)
+    @Operation(
+            summary = "Add cart item to the shopping cart",
+            description = "Add cart item to the shopping cart"
+    )
+    public CartItemDto addToCart(
+            Authentication authentication,
+            @RequestBody CreateCartItemDto requestDto
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/controller/CartController.java]:
```suggestion
            @RequestBody @Valid CreateCartItemDto requestDto
```

File: src/main/java/org/example/intro/dto/cart/CreateCartItemDto.java [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+import jakarta.validation.constraints.NotBlank;
+import jakarta.validation.constraints.Positive;
+import jakarta.validation.constraints.Size;
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+public class CreateCartItemDto {
+    @NotBlank(message = "The book identifier is mandatory")
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/dto/cart/CreateCartItemDto.java]:
```suggestion
    @NotBlank(message = "The book identifier is mandatory")
    @NotNull
```

File: src/main/java/org/example/intro/dto/cart/UpdateCartItemsQuantityDto.java [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+import jakarta.validation.constraints.NotBlank;
+import jakarta.validation.constraints.Positive;
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+public class UpdateCartItemsQuantityDto {
+    @NotBlank(message = "Quantity is mandatory field")
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/dto/cart/UpdateCartItemsQuantityDto.java]:
```suggestion
```

File: src/main/java/org/example/intro/mapper/CartItemMapper.java [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+            CreateCartItemDto requestDto
+    ){
+        cartItem.setBook(new Book(requestDto.getBookId()));
+    }
+
+    @Mapping(target = "bookId", source = "book.id")
+    @Mapping(
+            target = "bookTitle",
+            source = "book.title"
+    )
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/mapper/CartItemMapper.java]:
```suggestion
    @Mapping(target = "bookId", source = "book.id")
    @Mapping(target = "bookTitle", source = "book.title")
```

File: src/main/java/org/example/intro/model/CartItem.java [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+@Setter
+@Entity
+@Table(name = "cart_items")
+public class CartItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @EqualsAndHashCode.Exclude
+    @ToString.Exclude
+    @ManyToOne
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/model/CartItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/org/example/intro/model/CartItem.java [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    @JoinColumn(name = "shopping_cart_id", nullable = false)
+    private ShoppingCart shoppingCart;
+    @EqualsAndHashCode.Exclude
+    @ToString.Exclude
+    @ManyToOne
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/model/CartItem.java]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/org/example/intro/model/CartItem.java [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    @JoinColumn(name = "book_id", nullable = false)
+    private Book book;
+    @Column(nullable = false)
+    private int quantity = 1;
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/model/CartItem.java]:
```suggestion
    private int quantity;
```

File: src/main/java/org/example/intro/model/ShoppingCart.java [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+@NoArgsConstructor
+public class ShoppingCart {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @JoinColumn(name = "user_id", nullable = false)
+    @MapsId
+    @OneToOne(fetch = FetchType.LAZY)
+    private User user;
+    @OneToMany(mappedBy = "shoppingCart")
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/model/ShoppingCart.java]:
```suggestion
    @OneToMany(mappedBy = "shoppingCart", cascade = CascadeType.ALL, orphanRemoval = true)
```

File: src/main/java/org/example/intro/model/User.java [hunk 13]
```diff
@@ -53,6 +53,8 @@ public class User implements UserDetails {
     )
     @Column(nullable = false)
     private Set<Role> roles;
+    @OneToOne(mappedBy = "user", cascade = CascadeType.ALL)
+    private ShoppingCart shoppingCart;
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/model/User.java]: remove

File: src/main/java/org/example/intro/service/impl/CartServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+import org.example.intro.sequrity.CustomUserDetailsService;
+import org.example.intro.service.CartService;
+import org.springframework.data.domain.Page;
+import org.springframework.security.core.Authentication;
+import org.springframework.stereotype.Service;
+import org.springframework.data.domain.Pageable;
+
+@RequiredArgsConstructor
+@Service
+public class CartServiceImpl implements CartService {
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/service/impl/CartServiceImpl.java]: all logic with shop cart should be in ShoppingCartService only

File: src/main/java/org/example/intro/service/impl/UserServiceImpl.java [hunk 15]
```diff
@@ -30,7 +34,12 @@ public UserResponseDto register(UserRegistrationRequestDto requestDto)
...
         user.setPassword(passwordEncoder.encode(requestDto.getPassword()));
-        userRepository.save(user);
-        return userMapper.toUserResponseDto(user);
+        try {
+            userRepository.save(user);
+            cartService.createShoppingCart(user);
+            return userMapper.toUserResponseDto(user);
+        } catch (RuntimeException e) {
+            throw new RegistrationException(e.getMessage());
+        }
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/service/impl/UserServiceImpl.java]:
```suggestion
        userRepository.save(user);
        cartService.createShoppingCart(user);
        return userMapper.toUserResponseDto(user);
```

File: src/main/resources/db/changelog/changes/13-create-shopping-carts-table.yaml [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+            columns:
+              - column:
+                  name: id
+                  type: bigint
+                  autoIncrement: true
+                  constraints:
+                    primaryKey: true
+                    nullable: false
+              - column:
+                  name: user_id
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/13-create-shopping-carts-table.yaml]: as you use `@MapsId ` don't need to add this column

File: src/main/resources/db/changelog/changes/15-create-cart-items-table.yaml [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+                  name: id
+                  type: bigint
+                  autoIncrement: true
+                  constraints:
+                    primaryKey: true
+                    nullable: false
+              - column:
+                  name: shopping_cart_id
+                  type: bigint
+                  constraints:
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/15-create-cart-items-table.yaml:19]: foreignKeyName?

File: src/main/resources/db/changelog/changes/15-create-cart-items-table.yaml [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+                    nullable: false
+              - column:
+                  name: book_id
+                  type: bigint
+                  constraints:
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/15-create-cart-items-table.yaml:25]: foreignKeyName?

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
If you have some questions, please ask in chat

File: src/main/java/org/example/intro/controller/CartController.java (hunk 4 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/controller/CartController.java]: ++, thank you!

File: src/main/java/org/example/intro/controller/CartController.java (hunk 5 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/controller/CartController.java]: ++, thank you

File: src/main/java/org/example/intro/dto/cart/CreateCartItemDto.java (hunk 6 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/dto/cart/CreateCartItemDto.java]: ++, thanks

File: src/main/java/org/example/intro/dto/cart/UpdateCartItemsQuantityDto.java (hunk 7 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/dto/cart/UpdateCartItemsQuantityDto.java]: ++

File: src/main/java/org/example/intro/mapper/CartItemMapper.java (hunk 8 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/mapper/CartItemMapper.java]: ++

File: src/main/java/org/example/intro/model/CartItem.java (hunk 9 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/model/CartItem.java]: ++

File: src/main/java/org/example/intro/model/CartItem.java (hunk 10 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/model/CartItem.java]: ++

File: src/main/java/org/example/intro/model/CartItem.java (hunk 11 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/model/CartItem.java]: ++

File: src/main/java/org/example/intro/model/ShoppingCart.java (hunk 12 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/model/ShoppingCart.java]: ++

File: src/main/java/org/example/intro/model/User.java (hunk 13 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/model/User.java]: ++, removed

File: src/main/java/org/example/intro/service/impl/CartServiceImpl.java (hunk 14 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/service/impl/CartServiceImpl.java]: ++, renamed

File: src/main/java/org/example/intro/service/impl/UserServiceImpl.java (hunk 15 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/service/impl/UserServiceImpl.java]: ++, updated

File: src/main/resources/db/changelog/changes/15-create-cart-items-table.yaml (hunk 18 above)
Comment (artemklishch) [reply src/main/resources/db/changelog/changes/15-create-cart-items-table.yaml:25]: ++

File: src/main/resources/db/changelog/changes/15-create-cart-items-table.yaml (hunk 17 above)
Comment (artemklishch) [reply src/main/resources/db/changelog/changes/15-create-cart-items-table.yaml:19]: ++

File: src/main/resources/db/changelog/changes/13-create-shopping-carts-table.yaml (hunk 16 above)
Comment (artemklishch) [reply src/main/resources/db/changelog/changes/13-create-shopping-carts-table.yaml]: ++, thank you so much. updated

File: src/main/java/org/example/intro/controller/CartController.java (hunk 2 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/controller/CartController.java]: ++

File: src/main/java/org/example/intro/controller/CartController.java (hunk 3 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/controller/CartController.java]: ++

File: src/main/java/org/example/intro/dto/cart/CreateCartItemDto.java [hunk 19]
```diff
@@ -0,0 +1,18 @@
...
+import jakarta.validation.constraints.Positive;
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+public class CreateCartItemDto {
+    @NotNull(message = "The book identifier is mandatory")
+    private Long bookId;
+    @NotNull(message = "The quantity value is mandatory")
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/dto/cart/CreateCartItemDto.java]:
```suggestion
```

File: src/main/java/org/example/intro/service/impl/ShoppingCartServiceImpl.java [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+            Authentication authentication
+    ) {
+        Long bookId = createCartItemDto.getBookId();
+        Book bookById = bookRepository.findById(bookId)
+                .orElseThrow(() -> new NoSuchElementException("Book not found"));
+        Long userId = ((User) userDetailsService
+                .loadUserByUsername(authentication.getName())).getId();
+        CartItem existingCartItem = cartItemRepository.findByCartIdAndBookId(userId, bookId);
+        if (existingCartItem != null) {
+            throw new RuntimeException("You can not add more than one cart item " +
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/service/impl/ShoppingCartServiceImpl.java]: bad practice to throw RuntimeException, create custom one

File: src/main/java/org/example/intro/service/impl/ShoppingCartServiceImpl.java [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+    private final CartItemMapper cartItemMapper;
+    private final CartItemRepository cartItemRepository;
+    private final BookRepository bookRepository;
+
+    @Override
+    public ShoppingCartDto getCart(Authentication authentication) {
+        Long userId = ((User) userDetailsService
+                .loadUserByUsername(authentication.getName())).getId();
+        ShoppingCart cart = cartRepository.findById(userId)
+                .orElseThrow(() -> new NoSuchElementException("No cart found"));
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/service/impl/ShoppingCartServiceImpl.java]: add userId in message

File: src/main/resources/db/changelog/changes/13-create-shopping-carts-table.yaml [hunk 22]
```diff
@@ -0,0 +1,27 @@
...
+            tableName: shopping_carts
+            columns:
+              - column:
+                  name: id
+                  type: bigint
+                  autoIncrement: true
+                  constraints:
+                    primaryKey: true
+                    nullable: false
+#              - column:
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/13-create-shopping-carts-table.yaml]: remove comments

File: src/main/java/org/example/intro/controller/CartController.java [hunk 23]
```diff
@@ -0,0 +1,84 @@
...
+public class CartController {
+    private final ShoppingCartService cartService;
+
+    @GetMapping
+    @PreAuthorize("hasAnyRole('USER')")
+    @Operation(
+            summary = "Retrieve the shopping cart",
+            description = "Retrieve the shopping cart data with cart items"
+    )
+    public ShoppingCartDto getCart(Authentication authentication){
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/controller/CartController.java:39]:
```suggestion
    public ShoppingCartDto getCart(Authentication authentication){
         User user = (User) authentication.getPrincipal();
         return cartService.getCart(user.getId());
```

Comment (Elena-Bruyako) [reply src/main/java/org/example/intro/controller/CartController.java:39]: same for other endpoints

File: src/main/java/org/example/intro/service/impl/ShoppingCartServiceImpl.java [hunk 24]
```diff
@@ -0,0 +1,98 @@
...
+
+    @Override
+    public ShoppingCartDto updateCartItem(
+            Long cartItemId,
+            UpdateCartItemsQuantityDto requestDto,
+            Authentication authentication
+    ) {
+        CartItem cartItemFromDb = cartItemRepository
+                .findById(cartItemId)
+                .orElseThrow(() -> new NoSuchElementException("Cart item not found"));
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
                .orElseThrow(() -> new EntityNotFoundException("Cart item not found"));
```

File: src/main/java/org/example/intro/service/impl/ShoppingCartServiceImpl.java [hunk 25]
```diff
@@ -0,0 +1,98 @@
...
+        return getCart(authentication);
+    }
+
+    @Override
+    public ShoppingCartDto updateCartItem(
+            Long cartItemId,
+            UpdateCartItemsQuantityDto requestDto,
+            Authentication authentication
+    ) {
+        CartItem cartItemFromDb = cartItemRepository
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
        CartItem cartItem = cartItemRepository
```

File: src/main/java/org/example/intro/service/impl/ShoppingCartServiceImpl.java [hunk 26]
```diff
@@ -0,0 +1,98 @@
...
+    @Override
+    public void deleteCartItem(Long cartItemId, Authentication authentication) {
+        Long userId = ((User) userDetailsService
+                .loadUserByUsername(authentication.getName())).getId();
+        CartItem existingCartItem = Optional.ofNullable(cartItemRepository
+                .findByIdAndByCartId(userId, cartItemId)).orElseThrow(
+                () -> new NoSuchElementException(
+                        "No such item found in your shopping cart"
+                )
+        );
```
Comment (Elena-Bruyako) [inline src/main/java/org/example/intro/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
        CartItem existingCartItem = cartItemRepository
                .findByIdAndByCartId(userId, cartItemId).orElseThrow(
                () -> new EntityNotFoundException(
                        "No such item found in your shopping cart"
                )
        );
```


add parameters to message

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/org/example/intro/service/impl/ShoppingCartServiceImpl.java (hunk 26 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/service/impl/ShoppingCartServiceImpl.java]: ++, updated

File: src/main/java/org/example/intro/service/impl/ShoppingCartServiceImpl.java (hunk 25 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/service/impl/ShoppingCartServiceImpl.java]: ++, updated

File: src/main/java/org/example/intro/service/impl/ShoppingCartServiceImpl.java (hunk 24 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/service/impl/ShoppingCartServiceImpl.java]: ++, updated. Thank you!

File: src/main/java/org/example/intro/controller/CartController.java (hunk 23 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/controller/CartController.java:39]: ++, updated

File: src/main/resources/db/changelog/changes/13-create-shopping-carts-table.yaml (hunk 22 above)
Comment (artemklishch) [reply src/main/resources/db/changelog/changes/13-create-shopping-carts-table.yaml]: ++

File: src/main/java/org/example/intro/service/impl/ShoppingCartServiceImpl.java (hunk 21 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/service/impl/ShoppingCartServiceImpl.java]: ++

File: src/main/java/org/example/intro/service/impl/ShoppingCartServiceImpl.java (hunk 20 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/service/impl/ShoppingCartServiceImpl.java]: ++, updated. thank you.

File: src/main/java/org/example/intro/dto/cart/CreateCartItemDto.java (hunk 19 above)
Comment (artemklishch) [reply src/main/java/org/example/intro/dto/cart/CreateCartItemDto.java]: ++, updated.

**Pull request merged**